router/link/link_registry.go

```diff
@@ -325,13 +325,31 @@ func (self *linkRegistryImpl) evaluateLinkStateQueue() {
 }
 
 func (self *linkRegistryImpl) evaluateDestinations() {
-	for _, dest := range self.destinations {
-		// TODO: When do we drop destinations? Should we ask the controller after the router has been
-		//       unhealthy for a while and it doesn't have any established links? Do this on exponential backoff?
-		//      Should the controller send router removed messages?
+	for destId, dest := range self.destinations {
+		hasEstablishedLinks := false
 		for _, state := range dest.linkMap {
+			// verify that links marked as established have an open link. There's a small chance that a link established
+			// and link closed could be processed out of order if the event queue is full. This way, it will eventually
+			// get fixed.
+			if state.status == StatusEstablished {
+				link, _ := self.GetLink(state.linkKey)
+				if link == nil || link.IsClosed() {
+					state.retryDelay = time.Duration(0)
+					state.nextDial = time.Now()
+				} else {
+					hasEstablishedLinks = true
+				}
+			}
+
 			self.evaluateLinkState(state)
 		}
+
+		// we are notified of deleted routers. In case we're unreachable while a router is deleted,
+		// we will also stop trying to contact unhealthy routers after a period. If a destination
+		// has nothing to dial, it should also be removed
+		if len(dest.linkMap) == 0 || (!dest.healthy && !hasEstablishedLinks && time.Since(dest.unhealthyAt) > 8*time.Hour) {
+			delete(self.destinations, destId)
```

Comment:
this feels like a log-worthy event to me.  also 8 hours? Why 8 hours? Why not forever? It's also not configurable, could it glom onto some other setting that is?

Reply:
It's very low probability garbage collection. If we were notified of a router that had listeners, but then the router was later removed but we weren't available when the router delete notification was sent, this lets us GC the destination router. Or if a router is unhealthy and can't be reached for a long time we'll stop trying to reach it. Once it reconnected to the controller, we'll start trying to connect to it again.

I went back and forth on this. There's not a convenient config value to attach to. I could bump it to a few days. In generally if a router is not reach for a few hours, someone should get alerted and the problem should be fixed before this deadline is ever hit.

I'll bump it to 48 hours.